Mono.Debugging/Mono.Debugging.Client/DebuggerSession.cs

```diff
@@ -409,22 +409,20 @@ public void Run (DebuggerStartInfo startInfo, DebuggerSessionOptions options)
 				throw new ArgumentNullException (nameof (startInfo));
 			if (options == null)
 				throw new ArgumentNullException (nameof (options));
-			
-			lock (slock) {
-				this.options = options;
-				OnRunning ();
-				Dispatch (delegate {
-					try {
-						OnRun (startInfo);
-					} catch (Exception ex) {
-						// should handle exception before raising Exit event because HandleException may ignore exceptions in Exited state
-						var exceptionHandled = HandleException (ex);
-						ForceExit ();
-						if (!exceptionHandled)
-							throw;
-					}
-				});
-			}
+
+			Dispatch (delegate {
+				try {
+					this.options = options;
```

Comment:
The options assignment should be done before the Dispatch, since otherwise the Options and EvaluationOptions properties will not return the right value just after calling Run().

Reply:
If we need a thread-safe way to set it, we can use `Interlocked.Exchange(ref this.options, options)`

Reply:
And that way we also avoid the need for a lock for synchronization